File: cpp/sort/HEAP_SORT.cpp
#include<iostream>
#include<windows.h>

using namespace std;

void MAX_HEAPIFY(int *A, int length, int heap_size ,int i) {
	int l = 2 * i;
	int r = 2 * i + 1;
	int largest = 0;
	if (l <= heap_size && A[l-1] > A[i-1])
		largest = l;
	else
		largest = i;
	if (r <= heap_size&&A[r-1] > A[largest-1])
		largest = r;
	if (largest != i) {
		int temp = A[largest-1];
		A[largest-1] = A[i-1];
		A[i-1] = temp;
		MAX_HEAPIFY(A, length,heap_size,largest);
	}
}

void BUILD_MAX_HEAP(int *A, int length,int heap_size) {
	for (int i = length / 2; i > 0; i--)
		MAX_HEAPIFY(A, length, heap_size, i);
}

void HEAP_SORT(int *A, int length, int heap_size) {
	BUILD_MAX_HEAP(A, length, heap_size);
	for (int i = length; i > 1; i--) {
		int temp = A[0];
		A[0] = A[i-1];
		A[i-1] = temp;
		heap_size--;
		MAX_HEAPIFY(A, length, heap_size, 1);
	}
}
int main() {
	int A[] = { 4,1,3,2,16,9,10,14,8,7 };
	HEAP_SORT(A, sizeof(A) / sizeof(int), sizeof(A) / sizeof(int));
	for (int i = 0; i < sizeof(A) / sizeof(int); i++)
		cout << A[i] << " ";
	cout << endl;
	system("pause");
}

File: algorithm/sort/heap_sort.py
def left(i):
    return 2 * i + 1


def right(i):
    return 2 * i + 2


def parent(i):
    return i // 2


class Heap:
    def __init__(self, data):
        """
        length 表示数组元素的个数
        heap_size 表示有多少个堆元素存储在该数组中
        0 <= heap_size <= length
        """
        self._data = data
        self.length = len(self._data)
        self.heap_size = self.length

    def max_heapify(self, i):
        """ 假定根节点为 left(i) 和right(i)的二叉树都是最大堆，但是A[i]可能小于其孩子

        通过让A[i]在最大堆中逐级下降，从而使得以下标i为根节点的子树重新遵循最大堆的性质

        每个孩子的子树的大小至多为2n/3

        T(n) <= T(2n/3) + Theta(1)

        a = 1, b = 3/2, k = 0, 满足 a=b^k, 时间复杂度为O(log n)
        """
        l = left(i)
        r = right(i)

        if l < self.heap_size and self._data[l] > self._data[i]:
            largest = l
        else:
            largest = i

        if r < self.heap_size and self._data[r] > self._data[largest]:
            largest = r

        if largest != i:
            self._data[i], self._data[largest] = self._data[largest], self._data[i]
            self.max_heapify(largest)

    def build_max_heap(self):
        """
        叶节点不用调整
        时间复杂度为 O(n)
        """
        for i in range(self.length // 2, -1, -1):
            self.max_heapify(i)

    def heap_sort(self):
        """ 时间复杂度 O(n log n) ，n-1次调用max_heapify，每次 O(log n) """
        self.build_max_heap()
        for i in range(self.length-1, 0, -1):
            self._data[0], self._data[i] = self._data[i], self._data[0]
            self.heap_size -= 1
            self.max_heapify(0)

        return self._data


if __name__ == '__main__':
    A = [1,2,4,3,19,2,44,29,190,24]
    solution = Heap(A)
    print(solution.heap_sort())
